core/src/main/java/io/grpc/internal/InternalSubchannel.java

```diff
@@ -335,43 +293,43 @@ void resetConnectBackoff() {
         gotoNonErrorState(CONNECTING);
         startNewTransport();
       }
-    } finally {
-      syncContext.drain();
-    }
+    });
   }
 
-  @GuardedBy("lock")
-  private void gotoNonErrorState(ConnectivityState newState) {
-    gotoState(ConnectivityStateInfo.forNonError(newState));
+  private void gotoNonErrorState(final ConnectivityState newState) {
+    syncContext.execute(new Runnable() {
```

Comment:
`gotoNonErrorState()` is only run from synContext. Why is this necessary?

Reply:
iirc, this was hard to figure out. the SyncContext execution order is different than lock. because it doesn't allow the re-entrance.

Reply:
Did it cause any problem? Calling inline is closer to the previous behavior under lock.

Reply:
yes, 3 tests are failing (io.grpc.internal.ManagedChannelImplTest.{channelsAndSubchannels_oob_instrumented_fail,oobchannels,channelsAndSubchannels_oob_instrumented_success}.

Reply:
These tests fail because `transportReady()` delivers a `READY` state to the listener before updating  `activeTransport`. `OobChannel` calls `DelayedClientTransport.reprocess()` as soon as receiving a `READY`, which will fail because `activeTransport` is not yet set. Previous code got away with this issue because it called `syncContext.drain()` after `activeTransport` is set, but the root cause, which is transitioning to a new state prematurely, was already there. Using `syncContext.excute()` inside `gotoNonErrorState()` only covers up the root cause.

The correct implementation is to transition to a new state **after** all necessary state changes have been made. If you move all `gotoNonErrorState()` and `gotoState()` calls to after their prerequisite state changes, those tests will pass.